homeassistant/components/youtube/api.py

```diff
@@ -0,0 +1,41 @@
+"""API for YouTube bound to Home Assistant OAuth."""
+from aiohttp import ClientSession
+from google.auth.exceptions import RefreshError
+from google.oauth2.credentials import Credentials
+from google.oauth2.utils import OAuthClientAuthHandler
+from googleapiclient.discovery import Resource, build
+
+from homeassistant.const import CONF_ACCESS_TOKEN
+from homeassistant.helpers import config_entry_oauth2_flow
+
+
+class AsyncConfigEntryAuth(OAuthClientAuthHandler):
+    """Provide Google authentication tied to an OAuth2 based config entry."""
+
+    def __init__(
+        self,
+        websession: ClientSession,
+        oauth2_session: config_entry_oauth2_flow.OAuth2Session,
+    ) -> None:
+        """Initialize YouTube Auth."""
+        self.oauth_session = oauth2_session
+        super().__init__(websession)
+
+    @property
+    def access_token(self) -> str:
+        """Return the access token."""
+        return self.oauth_session.token[CONF_ACCESS_TOKEN]
+
+    async def check_and_refresh_token(self) -> str:
+        """Check the token."""
+        await self.oauth_session.async_ensure_token_valid()
+        return self.access_token
+
+    async def get_resource(self) -> Resource:
+        """Get current resource."""
+        try:
+            credentials = Credentials(await self.check_and_refresh_token())
+        except RefreshError as ex:
```

Comment:
Could it be that it was left behind from a bigger OAuth implementation since we now only declare an AuthorizationServer?

Reply:
> @allenporter can't react on my phone to the comment about reauth:
> 
> Eventually I got the tests working, instead of forcing a time update I forced the coordinator to update.
> 
> I do have a hard time trying to figure what's needed for a correctly working oauth setup and what's part of reauth since all implementations with oauth already have reauth in them so I don't know.

What you have seems like a good mix to me.